## 异常

![image-20200307225253438](C:\Users\86159\AppData\Roaming\Typora\typora-user-images\image-20200307225253438.png)

### Error和Exception

1. 概念：Exception和Error都是继承于Throwable 类，在 Java 中只有 Throwable 类型的实例才可以被抛出（throw）或者捕获（catch），它是异常处理机制的基本组成类型。
2. Exception是java程序运行中可预料的异常情况，咱们可以获取到这种异常，并且对这种异常进行业务外的处理。
3. Error是java程序运行中不可预料的异常情况，这种异常发生以后，会直接导致JVM不可处理或者不可恢复的情况。所以这种异常不可能抓取到，比如OutOfMemoryError、NoClassDefFoundError等。
4. Exception又分为运行时异常和非运行时异常。两个根本的区别在于，运行时异常必须在编写代码时，使用try catch捕获（比如：IOException异常）。非运行时异常在代码编写时，可以忽略捕获操作（比如：ArrayIndexOutOfBoundsException），这种异常是在代码编写或者使用过程中通过规范可以避免发生的。
   

![image-20200313141038857](C:\Users\86159\AppData\Roaming\Typora\typora-user-images\image-20200313141038857.png)

![image-20200313141748277](C:\Users\86159\AppData\Roaming\Typora\typora-user-images\image-20200313141748277.png)

* 一般在finally中关闭资源



![image-20200322003501772](C:\Users\86159\AppData\Roaming\Typora\typora-user-images\image-20200322003501772.png)

### JDK新特性

try-with-resource





![image-20200322004111447](C:\Users\86159\AppData\Roaming\Typora\typora-user-images\image-20200322004111447.png)